worlds/doom_1993/__init__.py

```diff
@@ -0,0 +1,296 @@
+        item_id: int = self.item_name_to_id[name]
+        doom_1993_item = DOOM1993Item(name, Items.item_table[item_id]["classification"], item_id, player=self.player)
+        return doom_1993_item
+    
+    
+    def create_item_with_classification(self, name: str, classification: ItemClassification) -> DOOM1993Item:
+        item_id: int = self.item_name_to_id[name]
+        doom_1993_item = DOOM1993Item(name, classification, item_id, player=self.player)
+        return doom_1993_item
+    
+
+    def place_locked_item_in_locations(self, item_name, locations):
+        location = self.multiworld.random.choice(locations)
+        self.multiworld.get_location(location, self.player).place_locked_item(self.create_item(item_name))
+        self.location_count -= 1
+
+
+    def create_items(self):
+        is_only_first_episode = self.get_episode_count() == 1 and self.included_episodes[0]
+
+        # Items
+        for item_id, item in Items.item_table.items():
+            if item["name"] == "Shotgun":
+                tmp = 5
+            if item["episode"] != -1 and not self.included_episodes[item["episode"] - 1]:
+                continue
+            if (item["name"] == "BFG9000" or item["name"] == "Plasma Gun") and is_only_first_episode:
+                continue # Don't include those guns in first episode
+            count = item["count"]
+            if str(item["name"]) in self.starting_level_for_episode:
+                count -= 1
+            for i in range(0, count):
+                classification = item["classification"]
+                if i < item["progression_count"]:
+                    classification = ItemClassification.progression
+                doom_1993_item = self.create_item_with_classification(item["name"], classification)
+                # See create_regions
+                if item["name"] != "Warrens (E3M9) - Blue skull key" and \
+                   item["name"] != "Halls of the Damned (E2M6) - Yellow skull key":
+                    self.multiworld.itempool += [doom_1993_item]
+                    self.item_count += 1
+
+        # Place end level items in locked locations
+        for event in Events.events:
+            if not event in self.location_name_to_id:
+                continue
+            loc = Locations.location_table[self.location_name_to_id[event]]
+            if not self.included_episodes[loc["episode"] - 1]:
+                continue
+            self.multiworld.get_location(event, self.player).place_locked_item(
+                DOOM1993Item(event, ItemClassification.progression, None, player=self.player))
+            self.location_count -= 1
+    
+        # Special case for E2M6 and E3M8, where you enter a normal door then get stuck behind with a key door.
+        # We need to put the key in the locations available behind this door.
+        if self.included_episodes[1]:
+            self.place_locked_item_in_locations("Halls of the Damned (E2M6) - Yellow skull key", [
```

Comment:
I feel like, conceptually, the random placement of items that need special handling belongs in `generate_basic`. Though it looks like you rely on counting the number of items you place here to add an accurate number of filler items.

Could someone who knows more comment on how "proper" this is?

Reply:
If I recall, I discussed this in the dev channel already (Or here), used to be in elsewhere and was told to put there. I could be wrong.

Reply:
Looks like it was in reference to modifying the item pool in `generate_basic`/`early`.

Items shouldn't be added to or removed from the itempool after `create_items`, so it's correct to have that code in here.

But I'm wondering whether it's proper to be placing items in locations during `create_items` instead of `generate_basic`. It may be fine, it just looks out of place to me.

Reply:
Hi @Zunawe , see comment above by @Jarno458 :

> Ideally you dont place any items outside of the def create_items(self) -> None: step as of https://github.com/ArchipelagoMW/Archipelago/pull/1460

This is why I moved them there

Reply:
Understood. It is correct to modify the itempool in `create_items`.

I'm not sure whether it is correct/proper to call `place_locked_item_in_locations` or otherwise place items randomly outside of functions like `generate_basic`, `pre_fill`, or similar.